src/knowledgegraph/construction/Pipeline.cpp

```diff
@@ -0,0 +1,1052 @@
+                            json obj = {{"source", source},
+                                        {"destination", destination},
+                                        {"properties", jsonEdge["properties"]}};
+
+                            if (sourceIndex == destIndex) {
+                                kg_pipeline_stream_handler_logger.debug("Thread " + std::to_string(i) +
+                                                                        " adding local edge to partition " +
+                                                                        std::to_string(sourceIndex));
+                                partitioner.addLocalEdge(obj.dump(), sourceIndex);
+                            } else {
+                                kg_pipeline_stream_handler_logger.debug("Thread " + std::to_string(i) +
+                                                                        " adding edge cut to partition " +
+                                                                        std::to_string(sourceIndex));
+                                partitioner.addEdgeCut(obj.dump(), sourceIndex);
+                            }
+                        } else {
+                            kg_pipeline_stream_handler_logger.error("Malformed line: missing source/destination ID: " +
+                                                                    line);
+                        }
+                    } catch (const json::parse_error& e) {
+                        kg_pipeline_stream_handler_logger.error("JSON parse error: " + std::string(e.what()) +
+                                                                " | Line: " + line);
+                    } catch (const std::invalid_argument& e) {
+                        kg_pipeline_stream_handler_logger.error("Invalid node ID (not an integer) in line: " + line);
+                    } catch (const std::out_of_range& e) {
+                        kg_pipeline_stream_handler_logger.error("Node ID out of range in line: " + line);
+                    } catch (const std::exception& e) {
+                        kg_pipeline_stream_handler_logger.error("Unexpected exception in line: " +
+                                                                std::string(e.what()));
+                    }
+                }
+                // else if (!isReading) {
+                //     kg_pipeline_stream_handler_logger.info("Thread " +
+                //     std::to_string(i) + " exiting due to isReading=false"); break;
+                // }
+            }
+            kg_pipeline_stream_handler_logger.info("Tuple thread finished for partition " + std::to_string(i));
+        });
+    }
+    for (auto& t : tupleThreads) {
+        t.join();
+    }
+
+
+    auto endTime = high_resolution_clock::now();
+    std::chrono::duration<double> duration = endTime - startTime;
+    kg_pipeline_stream_handler_logger.info("processTupleAndSaveInPartition completed in " +
```

Comment:
This kind of details we need to publish in open telemetry and vizualize in Grafana